ApplicationLibCode/ProjectDataModel/Flow/RimWellConnectivityTable.cpp

```diff
@@ -0,0 +1,1256 @@
+    {
+        return;
+    }
+
+    m_selectedTimeStep = m_case->timeStepDates().front();
+}
+
+//--------------------------------------------------------------------------------------------------
+/// Update selected "From Time Step" and "To Time Step" according to selected case.
+/// If both selected time steps exist for case, keep as is, otherwise select first and last time
+/// step in case.
+//--------------------------------------------------------------------------------------------------
+void RimWellConnectivityTable::setValidTimeStepRangeForCase()
+{
+    if ( !m_case || m_case->timeStepDates().empty() )
+    {
+        return;
+    }
+    if ( m_selectedFromTimeStep().isValid() && isTimeStepInCase( m_selectedFromTimeStep() ) && m_selectedToTimeStep().isValid() &&
+         isTimeStepInCase( m_selectedToTimeStep() ) )
+    {
+        return;
+    }
+
+    m_selectedFromTimeStep = m_case->timeStepDates().front();
+    m_selectedToTimeStep   = m_case->timeStepDates().back();
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+bool RimWellConnectivityTable::isTimeStepInCase( const QDateTime& timeStep ) const
+{
+    if ( !m_case ) return false;
+
+    return std::find( m_case->timeStepDates().cbegin(), m_case->timeStepDates().cend(), timeStep ) != m_case->timeStepDates().cend();
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+int RimWellConnectivityTable::getTimeStepIndex( const QDateTime timeStep, const std::vector<QDateTime> timeSteps ) const
+{
+    auto itr = std::find( timeSteps.begin(), timeSteps.end(), timeStep );
+    if ( itr == timeSteps.end() ) return -1;
+    return static_cast<int>( std::distance( timeSteps.begin(), itr ) );
+}
+
+//--------------------------------------------------------------------------------------------------
+///
+//--------------------------------------------------------------------------------------------------
+void RimWellConnectivityTable::setSelectedProducersAndInjectorsForSingleTimeStep()
+{
+    if ( !m_case || !m_flowDiagSolution || !m_selectedTimeStep().isValid() ) return;
+
+    const int timeStepIndex = getTimeStepIndex( m_selectedTimeStep, m_case->timeStepDates() );
+    if ( timeStepIndex == -1 )
+    {
+        m_selectedProducerTracersUiField = std::vector<QString>();
+        m_selectedInjectorTracersUiField = std::vector<QString>();
+    }
+
+    const std::vector<QString> producerVec = RimFlowDiagnosticsTools::producerTracersInTimeStep( m_flowDiagSolution, timeStepIndex );
+    std::vector<QString>       injectorVec = RimFlowDiagnosticsTools::injectorTracersInTimeStep( m_flowDiagSolution, timeStepIndex );
```

Comment:
Missing const?

Reply:
No, performing `push_back` on next line for appending `Reservoir` as injector tracer